src/librbd/cache/ParentCacheObjectDispatch.cc

```diff
@@ -0,0 +1,234 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/WorkQueue.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/Journal.h"
+#include "librbd/Utils.h"
+#include "librbd/io/ObjectDispatchSpec.h"
+#include "librbd/io/ObjectDispatcher.h"
+#include "librbd/io/Utils.h"
+#include "librbd/cache/ParentCacheObjectDispatch.h"
+#include "osd/osd_types.h"
+#include "osdc/WritebackHandler.h"
+
+#include <vector>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::cache::ParentCacheObjectDispatch: " \
+                           << this << " " << __func__ << ": "
+
+using namespace ceph::immutable_obj_cache;
+using librbd::util::data_object_name;
+
+namespace librbd {
+namespace cache {
+
+template <typename I>
+ParentCacheObjectDispatch<I>::ParentCacheObjectDispatch(
+    I* image_ctx) : m_image_ctx(image_ctx), m_cache_client(nullptr),
+    m_object_store(nullptr), m_initialized(false), m_re_connecting(false) {
+  std::string controller_path =
+    ((CephContext*)(m_image_ctx->cct))->_conf.get_val<std::string>("immutable_object_cache_sock");
+  m_cache_client = new CacheClient(controller_path.c_str(), m_image_ctx->cct);
+}
+
+template <typename I>
+ParentCacheObjectDispatch<I>::~ParentCacheObjectDispatch() {
+    delete m_object_store;
+    delete m_cache_client;
+}
+
+template <typename I>
+void ParentCacheObjectDispatch<I>::init() {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 5) << dendl;
+
+  if (m_image_ctx->parent != nullptr) {
+    ldout(cct, 5) << "child image: skipping" << dendl;
+    return;
+  }
+
+  C_SaferCond* cond = new C_SaferCond();
```

Comment:
Nit: make this method asynchronous if it's going to potentially block. Just pass a ```Context* on_finish``` as a method parameter and add a new (optional) state to the ```OpenImageRequest``` state machine

Reply:
Currently, i have modified this connection action to be asynchronous. If it fail to connect to RO daemon, will trigger re-connect at every read until session is created successfully. 

i am not sure whether i totally understand this comment.  For adding a new(optional) state to the OpenImageRequest, do you mean that two new and independent methods, e.g. OpenRequest::send_parent_cache() and OpenRequest::handle_parent_cache(), should be  added into librbd/image/OpenRequets file ?

Reply:
> Nit: make this method asynchronous if it's going to potentially block. Just pass a `Context* on_finish` as a method parameter and add a new (optional) state to the `OpenImageRequest` state machine

Exactly, yes (and update the little drawing of state machine transitions in the header file).

Reply:
done. i have added send_parent_cache and handle_parent_cache to OpenImageRequest file. Also update the drawing of state machine transitions. thanks